compiler/analysis_common.go

```diff
@@ -0,0 +1,366 @@
+package compiler
+
+import (
+	"crypto/sha1"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/antonmedv/expr/ast"
+	"github.com/antonmedv/expr/file"
+)
+
+func (c *compiler) analyzeCommonExpr(node ast.Node) {
+	switch n := node.(type) {
+	case *ast.NilNode:
+		c.analyzeCommonNilNode(n)
+	case *ast.IdentifierNode:
+		c.analyzeCommonIdentifierNode(n)
+	case *ast.IntegerNode:
+		c.analyzeCommonIntegerNode(n)
+	case *ast.FloatNode:
+		c.analyzeCommonFloatNode(n)
+	case *ast.BoolNode:
+		c.analyzeCommonBoolNode(n)
+	case *ast.StringNode:
+		c.analyzeCommonStringNode(n)
+	case *ast.ConstantNode:
+		c.analyzeCommonConstantNode(n)
+	case *ast.UnaryNode:
+		c.analyzeCommonUnaryNode(n)
+	case *ast.BinaryNode:
+		c.analyzeCommonBinaryNode(n)
+	case *ast.ChainNode:
+		c.analyzeCommonChainNode(n)
+	case *ast.MemberNode:
+		c.analyzeCommonMemberNode(n)
+	case *ast.SliceNode:
+		c.analyzeCommonSliceNode(n)
+	case *ast.CallNode:
+		c.checkCallNode(n)
+	case *ast.BuiltinNode:
+		c.analyzeCommonBuiltinNode(n)
+	case *ast.ClosureNode:
+		c.analyzeCommonClosureNode(n)
+	case *ast.PointerNode:
+		c.analyzeCommonPointerNode(n)
+	case *ast.ConditionalNode:
+		c.analyzeCommonConditionalNode(n)
+	case *ast.ArrayNode:
+		c.analyzeCommonArrayNode(n)
+	case *ast.MapNode:
+		c.analyzeCommonMapNode(n)
+	case *ast.PairNode:
+		// do nothing
+	default:
+		panic(fmt.Sprintf("undefined node type (%T)", node))
+	}
+}
+
+func (c *compiler) analyzeCommonNilNode(n *ast.NilNode) {
+	n.SetSubExpr("nil")
+}
+
+func (c *compiler) analyzeCommonIdentifierNode(n *ast.IdentifierNode) {
+	n.SetSubExpr(n.Value)
+}
+
+func (c *compiler) analyzeCommonIntegerNode(n *ast.IntegerNode) {
+	n.SetSubExpr(strconv.FormatInt(int64(n.Value), 10))
+}
+
+func (c *compiler) analyzeCommonFloatNode(n *ast.FloatNode) {
+	n.SetSubExpr(strconv.FormatFloat(n.Value, 'f', 10, 64))
+}
+
+func (c *compiler) analyzeCommonBoolNode(n *ast.BoolNode) {
+	n.SetSubExpr(strconv.FormatBool(n.Value))
+}
+
+func (c *compiler) analyzeCommonStringNode(n *ast.StringNode) {
+	n.SetSubExpr(strconv.Quote(n.Value))
+}
+
+func (c *compiler) analyzeCommonConstantNode(n *ast.ConstantNode) {
```

Comment:
emm, you meaning `String()` func in golang?

Reply:
Yes

Reply:
> Yes
At the beginning, I also wanted to use the interface of `String`, but when I used it later, I found that I still needed an interface to set the expression for checking common sub-expr. because of the expression optimzation in package `optimizer`, I can't set it at parse expr. meanwhile,  I found that there are function to set/get `file.Location`  in `ast.Node`. so I added GetSubExpr/SetSubExpr similarly.